File: leetcode/63-unique-paths-ii.py
# https://leetcode.com/problems/unique-paths-ii

from typing import List

class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        """ top-down approach (memoization) """
        if obstacleGrid[-1][-1] == 1:
            return 0
        
        ROWS, COLS = len(obstacleGrid), len(obstacleGrid[0])
        cache = dict()

        def dfs(r, c):
            if (r, c) in cache:
                return cache[(r, c)]
            if r == ROWS or c == COLS or obstacleGrid[r][c] == 1:
                return 0
            if r == ROWS - 1 and c == COLS - 1:
                return 1
            
            unique_paths = dfs(r, c+1) + dfs(r+1, c)
            cache[(r, c)] = unique_paths
            return unique_paths

        return dfs(0, 0)



    def unique_paths_with_obstacles(self, obstacleGrid: List[List[int]]) -> int:
        """ bottom-up approach (tabulation) """
        if obstacleGrid[-1][-1] == 1:
            return 0
        ROWS, COLS = len(obstacleGrid), len(obstacleGrid[0])
        dp = [0] * COLS
        dp[-1] = 1

        for r in reversed(range(ROWS)):
            for c in reversed(range(COLS)):
                if obstacleGrid[r][c] == 1:
                    dp[c] = 0
                elif c < COLS-1:
                    dp[c] += dp[c+1]

        return dp[0]

File: leetcode/63-unique-paths-ii.js
// https://leetcode.com/problems/unique-paths-ii

/** bottom-up approach (tabulation)
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {

  const m = obstacleGrid.length;
  const n = obstacleGrid[0].length;
  
  // base case, the bottom-right corner is an obstacle
  if (obstacleGrid[m-1][n-1] === 1) return 0;


  obstacleGrid[m-1][n-1] = 1;

  for (let r = m - 1; r >= 0; r -= 1) {
    for (let c = n - 1; c >= 0; c -= 1) {
      
      if (r === m-1 && c === n-1) continue;

      if (obstacleGrid[r][c] === 1) {
        obstacleGrid[r][c] = 0;
      } else {
        obstacleGrid[r][c] = (r+1 === m ? 0 : obstacleGrid[r+1][c]) + (c+1 === n ? 0 : obstacleGrid[r][c+1]);
      }
    }
  }

  return obstacleGrid[0][0];
  
};


/** top-down approach (memoization)
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles2 = function(obstacleGrid) {

  const m = obstacleGrid.length;
  const n = obstacleGrid[0].length;
  
  // base case, the bottom-right corner is an obstacle
  if (obstacleGrid[m-1][n-1] === 1) return 0;

  const memo = {};
  const dfs = (r = 0, c = 0) => {
    if (r === m || c === n || obstacleGrid[r][c] === 1) {
      return 0;
    }

    if (r === m-1 && c === n-1) {
      return 1;
    }

    const key = r + ',' + c;

    if (memo.hasOwnProperty(key)) {
      return memo[key];
    }

    const pathNums = dfs(r+1, c) + dfs(r, c+1);
    memo[key] = pathNums;

    return pathNums;
  }

  return dfs();

};
